cl_dll/hud.cpp

```diff
@@ -937,3 +939,22 @@ float CHud::GetSensitivity( void )
 	return m_flMouseSensitivity;
 }
 
+bool CHud::IsTriggerForSinglePlayer(color24 rendercolor)
+{
+	auto r = rendercolor.r;
+	auto g = rendercolor.g;
+	auto b = rendercolor.b;
+
+	if ((r == 128) && (g == 128) && (b == 128)) // trigger_autosave
```

Comment:
It's pretty brittle comparing by color values like this, because now you're locked out of easily changing the colors if needed (need to change on the server + here). Is it possible to compare the classname instead?

Reply:
> It's pretty brittle comparing by color values like this, because now you're locked out of easily changing the colors if needed (need to change on the server + here). Is it possible to compare the classname instead?

There is only one method to get a classname on the client-side that I can think off, but believe me, this method will eat up performance even more and its implementation is a bit of a headache, but let me describe this concept to you briefly:

## First step
Get list of all entities on map by reading .bsp file
Or with `cl_entity_t *worldmodel = gEngfuncs.GetEntityByIndex(0);` and `worldmodel->model->entities`

## Second step

Let's take for example some entity from list for analyze:

```
{
"model" "*63"
"target" "tp_01"
"classname" "trigger_teleport"
}
```

In our case, we will be interested in only two keyvalues: `"classname"` and `"model"`

You may ask, why do we need the model name? Well, that is exactly the key to how we will differ them from other entities! Because the each brush entity has a unique model for it as mandatory!

# Third step

https://github.com/YaLTeR/OpenAG/blob/0ad666a1f2bee01da47e2880877d6d3c735be06b/cl_dll/hud_spectator.cpp#L237

Take the code from the `UTIL_FindEntityInMap` function as a basis and rewrite it to suit your needs, execute it only once (when changing the map) and write the results to the some `std::vector` variables

# Fourth step

```cpp
			for (int e = 0; e < MAX_EDICTS; ++e)
			{
				cl_entity_t* ent = gEngfuncs.GetEntityByIndex(e);
				if (ent && ent->model)
				{
					if (!entities_classname.empty() && !entities_model.empty())
					{
						for (size_t i = 0; i < entities_classname.size(); i++)
						{
							if (!strcmp(entities_model[i].c_str(), ent->model->name))
							{
								// Rest code
							}
						}
					}
				}
			}
```

Reply:
Therefore, I see the hardcoding colors on plugin side is rational way in that case, since there doesn't exist the straight way to get classname, only possible by the method that I described above

And besides, we can still change that colors on client-side, so that doesn't sound like issue at all

Reply:
Hmm, right. I don't see an obvious way to go from `cl_entity_t` to `entvars_t` (which has the classname).